precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+};
+use pallet_randomness::{
+	BalanceOf, GetBabeData, Pallet, Request, RequestInfo, RequestState, RequestType,
+};
+use precompile_utils::{costs::call_cost, prelude::*};
+use sp_core::{H160, H256, U256};
+use sp_std::{fmt::Debug, marker::PhantomData, vec, vec::Vec};
+
+// #[cfg(test)]
+// mod mock;
+// #[cfg(test)]
+// mod tests;
+mod solidity_types;
+use solidity_types::*;
+
+#[generate_function_selector]
+#[derive(Debug, PartialEq)]
+pub enum Action {
+	RelayEpochIndex = "relayEpochIndex()",
+	GetRequestStatus = "getRequestStatus(uint256)",
+	GetRequest = "getRequest(uint256)",
+	RequestRelayBabeEpochRandomWords =
+		"requestRelayBabeEpochRandomWords(address,uint256,uint64,bytes32,uint8)",
+	RequestLocalVRFRandomWords =
+		"requestLocalVRFRandomWords(address,uint256,uint64,bytes32,uint8,uint64)",
+	FulfillRequest = "fulfillRequest(uint256)",
+	IncreaseRequestFee = "increaseRequestFee(uint256,uint256)",
+	PurgeExpiredRequest = "purgeExpiredRequest(uint256)",
+}
+
+pub const FULFILLMENT_OVERHEAD_ESTIMATED_COST: u64 = 1000u64; // TODO: get real value from benchmarking
+pub const LOG_FULFILLMENT_SUCCEEDED: [u8; 32] = keccak256!("FulFillmentSucceeded()");
+pub const LOG_FULFILLMENT_FAILED: [u8; 32] = keccak256!("FulFillmentFailed()");
+
+pub fn log_fulfillment_succeeded(address: impl Into<H160>) -> Log {
+	log1(address, LOG_FULFILLMENT_SUCCEEDED, vec![])
+}
+
+pub fn log_fulfillment_failed(address: impl Into<H160>) -> Log {
+	log1(address, LOG_FULFILLMENT_FAILED, vec![])
+}
+
+/// Reverts if fees and gas_limit are not sufficient to make subcall and cleanup
+fn ensure_can_provide_randomness<Runtime>(
+	code_address: H160,
+	gas_limit: u64,
+	request_fee: BalanceOf<Runtime>,
+	clean_up_cost: u64,
+) -> EvmResult<()>
+where
+	Runtime: pallet_randomness::Config + pallet_evm::Config + pallet_base_fee::Config,
+	BalanceOf<Runtime>: Into<U256>,
+{
+	// assert fee > gasLimit * base_fee
+	let gas_limit_as_u256: U256 = gas_limit.into();
+	if let Some(gas_limit_times_base_fee) =
+		gas_limit_as_u256.checked_mul(pallet_base_fee::Pallet::<Runtime>::base_fee_per_gas())
+	{
+		if gas_limit_times_base_fee >= request_fee.into() {
+			return Err(revert(
+				"Gas limit at current price must be less than fees allotted",
+			));
+		}
+	} else {
+		return Err(revert("Gas limit times base fee overflowed U256"));
+	}
+	let log_cost = log_fulfillment_failed(code_address)
+		.compute_cost()
+		.map_err(|_| revert("failed to compute log cost"))?;
+	// Cost of the call itself that the batch precompile must pay.
+	let call_cost = call_cost(U256::zero(), <Runtime as pallet_evm::Config>::config());
+	// assert gasLimit > overhead cost
+	let overhead = call_cost + log_cost + clean_up_cost;
+	if gas_limit <= overhead {
+		return Err(revert("Gas limit must exceed overhead call cost"));
+	}
+	Ok(())
+}
+
+/// Subcall to provide randomness
+/// caller must call `ensure_can_provide_randomness` before calling this function
```

Comment:
This is problematic because `ensure_can_provide_randomness` will throw an error in some cases. Is it possible to circumvent these if the caller wants to?